src/block-walker.ts

```diff
@@ -0,0 +1,214 @@
+            }
+        }
+
+        // Helper functions
+        function get_prev_graph_node_and_attach_current_as_next(
+            prev_block_id: string, 
+            current_block_graph_node: TopShardChainBlockGraphNode | MidShardChainBlockGraphNode,
+            current_shardchain_block_ids: { id: string, depth: number }[],
+            current_depth: number,
+        ) {
+            let prev_graph_node = blocks.get(prev_block_id) as RootShardchainBlockGraphNode | MidShardChainBlockGraphNode;
+            if (prev_graph_node) {
+                if (!prev_graph_node.next) {
+                    // This case is for "Root" nodes
+                    prev_graph_node.next = current_block_graph_node;
+                } else if (!prev_graph_node.next_alt) {
+                    prev_graph_node.next_alt = current_block_graph_node;
+                } else {
+                    throw new Error(`Triple next block for block "${prev_graph_node.id}"`);
+                }
+            } else {
+                prev_graph_node = {
+                    type: "Mid",
+                    id: prev_block_id,
+                    next: current_block_graph_node,
+                };
+                blocks.set(prev_block_id, prev_graph_node);
+                current_shardchain_block_ids.push({ id: prev_block_id, depth: current_depth + 1 });
+            }
+            return prev_graph_node;
+        }
+    }
+
+    private async flatten_graph_and_prepare_result(): Promise<void> {
+        this.result = Array.from(this.blocks!.values())
+            .filter(f => f.type != "Root")
+            .sort((a, b) =>
+                a.gen_utime != b.gen_utime
+                ? a.gen_utime! - b.gen_utime!
+                : (a.workchain_id != b.workchain_id
+                ? a.workchain_id!.localeCompare(b.workchain_id!)
+                : a.shard!.localeCompare(b.shard!)))
+            .map(node => {
+                return {
+                    block_id: node.id,
+                    transaction_ids: node.transaction_ids,
+                    message_ids: node.message_ids,
+                };
+            })
+            .concat({
+                block_id: this.masterchain_block_info!._key,
+                transaction_ids: this.masterchain_block_info!.transaction_ids,
+                message_ids: this.masterchain_block_info!.message_ids,
+            })
+            .reduce((result, current) => {
+                result.block_ids.push(current.block_id);
+                result.message_ids.push(...current.message_ids!);
+                result.transaction_ids.push(...current.transaction_ids!);
+                return result;
+            }, {
+                block_ids: [] as string[],
+                transaction_ids: [] as string[],
+                message_ids: [] as string[],
+            });
+
+        if (RangerConfig.debug)
+            console.log(this.result);
+    }
+}
+
+export type RootShardchainBlockGraphNode = {
+    type: "Root",
```

Comment:
I suggest to introduce enum BlockGraphNodeType.

Reply:
Typescript has a good support for this case. ShardChainBlockGraphNode.type has a type "Root" | "Mid" | "Top" with a compile time errors for other strings.